### PR TITLE
Sh beacon

### PR DESCRIPTION
**Pull request by thatch45**
This beacon tracks all shells on the system and yields the commands executed in those shells into the event bus.

### CODE REVIEW

File: salt/modules/cmdmod.py [hunk 1]
```diff
@@ -1524,3 +1524,30 @@ def _is_valid_shell(shell):
...
+            for line in lines:
+                line = line.strip()
+                if line.startswith('#'):
+                    continue
+                elif not line:
+                    continue
+                else:
+                    ret.append(line)
+        except OSError:
+            return ret
```
Comment (isbm) [inline salt/modules/cmdmod.py]: No one would know what happened. Any logs here, perhaps?

File: salt/modules/cmdmod.py [hunk 2]
```diff
@@ -1524,3 +1524,30 @@ def _is_valid_shell(shell):
...
+    '''
+    Lists the valid shells on this system via the /etc/shells file
+
+    CLI Example::
+
+        salt '*' cmd.shells
+    '''
+    shells_fn = '/etc/shells'
+    ret = []
+    if os.path.exists(shells_fn):
```
Comment (isbm) [inline salt/modules/cmdmod.py]: I would just return here with an inverse condition, like:
```python
if not os.path.exists(shells_fn):
    log.error("File '{0}' was not found".format(shells_fn))
    return ret
```

File: salt/modules/cmdmod.py [hunk 3]
```diff
@@ -1524,3 +1524,30 @@ def _is_valid_shell(shell):
...
+            with salt.utils.fopen(shells_fn, 'r') as shell_fp:
+                lines = shell_fp.read().splitlines()
+            for line in lines:
+                line = line.strip()
+                if line.startswith('#'):
+                    continue
+                elif not line:
+                    continue
+                else:
+                    ret.append(line)
```
Comment (isbm) [inline salt/modules/cmdmod.py]: What's wrong with:
```python
try:
    ret = [line for line in salt.utils.fopen(shells_fn, "r").read().splitlines()
           if line and not line.startswith("#")]
except OSError:
    log.error(....
```

File: salt/modules/cmdmod.py [hunk 4]
```diff
@@ -1524,3 +1524,30 @@ def _is_valid_shell(shell):
...
+        except OSError:
+            return ret
+    return ret
```
Comment (isbm) [inline salt/modules/cmdmod.py]: Returning `ret` twice after exception handling is not very nice, IMO. Exception handling should do something to logging or re-raise crash, if needed.

File: salt/beacons/sh.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    ps_out = __salt__['status.procs']()
+    track_pids = []
+    for pid in ps_out:
+        if ps_out[pid].get('cmd', '') in shells:
+            track_pids.append(pid)
+    if pkey not in __context__:
+        __context__[pkey] = {}
+    for pid in track_pids:
+        if pid not in __context__[pkey]:
+            cmd = ['strace', '-f', '-e', 'execve', '-p', '{0}'.format(pid)]
```
Comment (isbm) [inline salt/beacons/sh.py]: Maybe some guard should exist, like "This system has no strace". Windows?

File: salt/modules/cmdmod.py (hunk 3 above)
Comment (thatch45) [inline salt/modules/cmdmod.py]: I don't like list comprehensions, I think this is more readable.

File: salt/modules/cmdmod.py (hunk 4 above)
Comment (thatch45) [inline salt/modules/cmdmod.py]: The solution here is simple, put the log error here that you suggest higher up and then just don't have the return in the block

File: salt/beacons/sh.py (hunk 5 above)
Comment (thatch45) [inline salt/beacons/sh.py]: Yes, this module still needs a virtual function

**Pull request merged**